keps/sig-instrumentation/1748-pod-resource-metrics/README.md

```diff
@@ -0,0 +1,431 @@
+
+* `kube_running_pod_resource_requests` contains series for the consumed requested resources for a pod at a given time. It will always report the `node` label since running pods must be assigned to a node.
+* `kube_running_pod_resource_limits` contains series for the consumed limit resources for a pod at a given time. It will always report the `node` label since running pods must be assigned to a node.
+
+This will add `O(pods * resource types)` metrics series, and in general most clusters have 4-8 resource types. The series per pod is `non_zero_resource_count * 2 * 2 (requests and limits for normal and _lifecycle)`.  This explicitly does not expose containers as a dimension, as the resource model applies rules to containers that cannot be trivially expressed via a formula and does not help answer the core questions.
+
+These metrics should roughly correspond to:
+
+* The resources the scheduler will consider the pod to be consuming at the current time
+* The resources the kubelet will use to decide whether to admit the pod (nodes will reject requested resources that exceed their actual allocatable capacity)
+* The effective reserved capacity as resource quota would observe for the namespace
+* The sum of all resources consumed on the cluster
+* The size of cgroup constraints applied to pods on a node
+
+The lifecycle metrics allow key insights that are impossible today:
+
+* Observing when init containers or containers have substantially different resource reservation, causing workloads to under provision nodes
+* Visualize batch compute consumption on the cluster where completed pods can be quickly excluded
+
+The metrics would be served by the scheduler that is holding the active scheduler lease. During a failover, no metrics would be reported. This would prevent duplicate metrics from being scraped and aligns with our decision in the scheduler to not fill caches prior to election for resource usage minimization.
+
+Custom schedulers based on the scheduler code should be able to easily reuse and report these metrics, and the configuration of the cluster instrumentation to retrieve those metrics would be left as an exercise for the integrator.
+
+
+### Add recording rules consistent with this metric to describe actual resource usage
+
+In order to make these metrics useful for calculation of actual usage, we would represent the current usage in a form that can easily be queried by applying recording rules in the standard Prometheus stack and equivalents in other systems without having to apply complex calculations.
+
+This section is normative and will be incrementally approached during alpha - not all resource metrics must appear in this format.
+
+* `kube_running_pod_resource_usage` - the resources currently consumed by a running pod
```

Comment:
can you clarify who is reporting this metric? the scheduler doesn't know the actual usage, so it has to be kubelet, right?

Reply:
This is the requested resources for running pods, so a subset of scheduled pods.

Reply:
how is this different from `kube_running_pod_resource_requests`?

Reply:
Oh, sorry, the scroll hid the metric.  Yes, this is data we already gather from kubelets in various sig-instrumentation projects.